hnn_core/neuron.py

```diff
@@ -0,0 +1,681 @@
+    neuron_net.aggregate_currents()
+    # combine net.current{} variables on each proc
+    _pc.allreduce(neuron_net.current['L5Pyr_soma'], 1)
+    _pc.allreduce(neuron_net.current['L2Pyr_soma'], 1)
+
+    _pc.barrier()  # get all nodes to this place before continuing
+
+    dpl_data = np.c_[np.array(dp_rec_L2.to_python()) +
+                     np.array(dp_rec_L5.to_python()),
+                     np.array(dp_rec_L2.to_python()),
+                     np.array(dp_rec_L5.to_python())]
+
+    dpl = Dipole(np.array(t_vec.to_python()), dpl_data)
+    if rank == 0:
+        if neuron_net.params['save_dpl']:
+            dpl.write('rawdpl.txt')
+
+        dpl.baseline_renormalize(neuron_net.params)
+        dpl.convert_fAm_to_nAm()
+        dpl.scale(neuron_net.params['dipole_scalefctr'])
+        dpl.smooth(neuron_net.params['dipole_smooth_win'] / h.dt)
+
+    neuron_net.trial_idx += 1
+
+    return dpl
+
+
+def load_custom_mechanisms():
+    import platform
+    import os.path as op
+
+    global _loaded_dll
+
+    if _loaded_dll is not None:
+        return
+
+    if platform.system() == 'Windows':
+        mech_fname = op.join(op.dirname(__file__), '..', 'mod', 'nrnmech.dll')
+    else:
+        mech_fname = op.join(op.dirname(__file__), '..', 'mod', 'x86_64',
+                             '.libs', 'libnrnmech.so')
+    h.nrn_load_dll(mech_fname)
+    _loaded_dll = mech_fname
+
+    if _get_rank() == 0:
+        print('Loading custom mechanism files from %s' % mech_fname)
+
+    return
+
+
+def _get_nhosts():
+    """Return the number of processors used by ParallelContext
+
+    Returns
+    -------
+    nhosts: int
+        Value from pc.nhost()
+    """
+    if _pc is not None:
+        return int(_pc.nhost())
+    else:
+        return 1
+
+
+def _get_rank():
+    """Return the MPI rank from ParallelContext
+
+    Returns
+    -------
+    rank: int
+        Value from pc.id()
+    """
+    if _pc is not None:
+        return int(_pc.id())
+    else:
+        return 0
+
+
+def _create_parallel_context(n_cores=None):
+    """Create parallel context.
+
+    Parameters
+    ----------
+    n_cores: int | None
+        Number of processors to use for a simulation. A value of None will
+        allow NEURON to use all available processors.
+    """
+
+    global _cvode, _pc
+
+    if _pc is None:
+        if n_cores is None:
+            # MPI: Initialize the ParallelContext class
+            _pc = h.ParallelContext()
+        else:
+            _pc = h.ParallelContext(n_cores)
+
+        _cvode = h.CVode()
+
+        # be explicit about using fixed step integration
+        _cvode.active(0)
+
+        # use cache_efficient mode for allocating elements in contiguous order
+        # cvode.cache_efficient(1)
+    else:
+        # ParallelContext() has already been called. Don't start more workers.
+        # Just tell old nrniv workers to quit.
+        _pc.done()
+
+
+class _neuron_network(object):
```

Comment:
camel case for class. I thought before that this was a function

Reply:
Fixed.